### PR TITLE
CRM-21050: Use datepicker for all date fields used in contribution backoffice form

### PR DESCRIPTION
**Pull request by pradpnayak**
Overview
----------------------------------------
Use datepicker for datefields used in backoffice contribtution, these are:
1. Recieve Date
2. Thank you Date
3. Receipt Date
4. Cancel/Refund Date

Before
----------------------------------------
Date fields were using jcalender.

After
----------------------------------------
Replaced all the date fields with datepicker. Here's the screencast to show that default date values and correctly set after submission:
test-multiple-after

---

 * [CRM-21050: Use datepicker for backoffice date fields](https://issues.civicrm.org/jira/browse/CRM-21050)
  

NOTE original PR was tackling this in a different way - hence comments don't appear to make sense.

### CODE REVIEW

**Review by monishdeb — APPROVED** [review]
Tested, working fine.

**Comment by eileenmcnaughton** [comment]
This makes me sad. We really want to switch to datepicker rather than deal with this nasty handling.

However, I'm ok for this to be merged after one clarification. It seems to me it is not receive_date (per the title of the PR) but rather trxn_date that is not being set correctly. Is that correct? If so can we correct that in the PR / JIRA?

**Comment by mattwire** [comment]
@eileenmcnaughton 
> This makes me sad. We really want to switch to datepicker rather than deal with this nasty handling.
Agreed, but I've abandoned datepicker so many times as I cannot, for want of trying, get it to set a default date passed in from CiviCRM.  With the old method this is easy!

**Comment by eileenmcnaughton** [comment]
is this actually adding trxn_date to the form? If so it seems a bigger change than I realised

**Comment by eileenmcnaughton** [comment]
@mattwire - this patch works on that form to convert thank you date & loads the default ok - does it offer you any clues as to the problem you were having?

https://github.com/eileenmcnaughton/civicrm-core/commit/d0177d88f1c78ec57af0162875d110a12476f7c4

Note that the Contribution.xml already has metadata to describe the date field so I didn't need to add it

**Comment by JoeMurray** [comment]
@eileenmcnaughton there are three date fields displayed on the form, this ensures that all three have their correct time displayed rather than 12:00am for the part of the form we're talking about is the payment section. The before and after images show the payment trxn_date now get their time displayed unlike before. receive_date processing should be unchanged.

**Comment by eileenmcnaughton** [comment]
Right - but I think if we fix any date fields we should change them to use the preferred datepicker at that point. We shouldn't fix the deprecated date field version - the patch above was the example of how to do that for one field

**Comment by colemanw** [comment]
@mattwire the confusing thing about switching to the new datpicker is that it uses a different format than the old method, which might mess up the defaults if you're dealing with legacy code. The old fields would use a localized format across 2 fields (one for date, one for time) which got complicated and messy. The new datepicker widget contains a standard ISO date/time value, which should be much easier to work with once you clear away the legacy stuff.

**Comment by monishdeb** [comment]
Jenkins test this please

**Comment by monishdeb** [comment]
@eileenmcnaughton @colemanw I have changed the JIRA title and description, because as per change it's an improvement to use datepicker for all date fields used in backoffice contribution. Also, I have shared a screencast above (on test build site) to show that now the date field values are correctly submitted and set as default on the edit form.

File: tests/phpunit/CRM/Activity/BAO/ActivityTest.php [hunk 1]
```diff
@@ -757,23 +757,18 @@ public function testbyActivityDateAndStatus() {
     $lastWeekActivities = array(1, 2, 3);
     $todayActivities = array(4, 5, 6, 7);
     $lastTwoMonthsActivities = array(8, 9, 10, 11);
-    $lastYearActivties = array(12, 13, 14, 15, 16);
```
Comment (eileenmcnaughton) [inline tests/phpunit/CRM/Activity/BAO/ActivityTest.php]: I think this should be in a separate PR - even if we have to ignore the test fail for now

File: tests/phpunit/CRM/Contribute/BAO/ContributionTest.php [hunk 2]
```diff
@@ -1203,8 +1203,6 @@ public function createContributionWithTax() {
     $form->testSubmit(array(
        'total_amount' => 100,
         'financial_type_id' => $financialType['id'],
-        'receive_date' => '04/21/2015',
```
Comment (eileenmcnaughton) [inline tests/phpunit/CRM/Contribute/BAO/ContributionTest.php]: can we update the format on these dates rather than remove?

**Comment by eileenmcnaughton** [comment]
Looks solid from a code POV - I haven't tested

File: tests/phpunit/CRM/Contribute/BAO/ContributionTest.php (hunk 2 above)
Comment (monishdeb) [reply tests/phpunit/CRM/Contribute/BAO/ContributionTest.php]: @eileenmcnaughton 2 reasons why I removed it:
1. Provided the default value of current date time in the desired format [here](https://github.com/civicrm/civicrm-core/pull/10841/files#diff-0573f78042d230ffc1d5e61c1749ae77R1291)
2. In these unit-tests receive date has no importance, however I have retained in some UT like [here](https://github.com/civicrm/civicrm-core/pull/10841/files#diff-4cf264b86743b01490484978a3bdd2d8R1690) where it's needed.

Or do you still want to update their formats?

Comment (eileenmcnaughton) [reply tests/phpunit/CRM/Contribute/BAO/ContributionTest.php]: that seems fine - sounds like we are not losing any test data

**Comment by monishdeb** [comment]
@eileenmcnaughton I have moved the additional UT failure fix in separate PR https://github.com/civicrm/civicrm-core/pull/11479

**Comment by monishdeb** [comment]
Jenkin test this please

**Comment by monishdeb** [comment]
Jenkin test this please

**Comment by monishdeb** [comment]
Jenkin test this please

**Comment by eileenmcnaughton** [comment]
test failure is unrelated

**Comment by colemanw** [comment]
@civicrm-builder retest this please

**Comment by colemanw** [comment]
I'm satisfied with the amount of review this has seen. Adding "merge  on pass" label.

**Comment by monishdeb** [comment]
Additional test build failure is not related and fixed here https://github.com/civicrm/civicrm-core/pull/11479

**Comment by eileenmcnaughton** [comment]
Merging in accordance with merge on pass as fail is unrelated

**Comment by monishdeb** [comment]
Thanks @eileenmcnaughton and @colemanw for your review and getting it merged :)

**Comment by jitendrapurohit** [comment]
Fatal error is displayed when a user tries to delete a contribution after this change. Raised a fix at https://github.com/civicrm/civicrm-core/pull/11500
